### PR TITLE
Fix `last_active_ts` Presence bug

### LINKED ISSUE #12424: last_active_ago is updated on sync when it shouldn't be
### Description

The `last_active_ts` is updated for a user when they perform a regular sync request where their previous presence is 'online'. It looks like synapse intends to only update this when a sync is performed with `set_presence = online` and the previous presence state is online, and the spec says it should only be updated on a proactive event from the user.

This is because `set_state` is called in order to set the new presence state and this method always updated last_active_ago.

Found while working on https://github.com/matrix-org/synapse/pull/12213 but I think this will need another inter-process API added to fix.

### Steps to reproduce

- Have two users, leave one syncing.
- On the other user, perform a sync request
- Notice on the first sync stream that the last_active_ago value for the other user reflects the the sync you just did.

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**:

If not matrix.org:

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**:

- **Install method**:
<!-- examples: package manager/git clone/pip  -->

- **Platform**:
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

### PR DESCRIPTION
**Pull request by realtyem**
* Fixes #12424 

Essentially, what was happening was that during an incremental sync, `online` is set repeatedly by the system(which is correct behavior, approximately each `timeout`). This means it was as simple as comparing the 'before' and 'after' state of `online`, if they matched then don't bump `last_active_ts`. I spent hours checking permutations, I'm pretty sure I caught them all but need more eyes to check me.

I did write a basic test to check the behavior(which fails on develop), but am open to ideas of what else to actually check for regression.

* [`last_active_ago`](https://spec.matrix.org/v1.7/client-server-api/#last-active-ago) in the Matrix Spec

Consequences:
* By not updating the timestamp every sync:
  * this avoids setting a staggered set of wheel timers. It appears they deduplicate into the same time bucket as was probably intended(5 seconds). Less processing of unnecessary timers is a small win.
  * this allows the `last_active_change_online` and `last_active_change_not_online` metrics to actually update correctly(which will be interesting to see how it plays out)
  * this may cause slightly more presence activity, as now users will be marked UNAVAILABLE(idle) correctly where they didn't before(it will probably be a wash, though).
* One less call to `_update_states()` machinery, which is the heavy lift of sending notifications and persisting changes(and setting more timers).

<details><summary>A quick hackish(and incomplete) primer on how presence does timeouts</summary>

* A local user that is online and currently syncing:
  * if they are currently sending messages, they have LAST_ACTIVE_GRANULARITY(60 seconds) before they are marked not 'currently_active'
  * if they are just sitting there and watching, they will be marked UNAVAILABLE(idle) in IDLE_TIMER(5 minutes) if they don't send a message or create a new 'pro-active' event(this was not happening correctly a lot of the time, however typing, sending read receipts, and sending read markers will bump last_active_ts, even though I don't believe they are expressly 'pro-active')
  * when they stop syncing:
    * in UPDATE_SYNCING_USERS_MS(10 seconds) we notify any workers that the user has stopped syncing
    * in SYNC_ONLINE_TIMEOUT(30 seconds) we assume the client is gone OFFLINE and mark them as such(unless they somehow create a 'pro-active' event). This ignores currently_active
* if the user is remote:
  * if we haven't heard from their server in FEDERATION_TIMEOUT(30 minutes), we mark them OFFLINE
* if the user is local but in a room with remote users:
  * we send presence every FEDERATION_PING_INTERVAL(25 minutes) so they don't mark us OFFLINE incorrectly.
</details>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Jason Little <realtyem@gmail.com>

### CODE REVIEW

File: synapse/handlers/presence.py [hunk 1]
```diff
@@ -1005,35 +1006,29 @@ async def user_syncing(
...
-            if prev_state.state == PresenceState.OFFLINE:
-                await self._update_states(
-                    [
-                        prev_state.copy_and_replace(
-                            state=PresenceState.ONLINE,
-                            last_active_ts=self.clock.time_msec(),
-                            last_user_sync_ts=self.clock.time_msec(),
-                        )
-                    ]
-                )
```
Comment (realtyem) [inline synapse/handlers/presence.py:1031]: This code block preformed two functions, updating `last_active_ts` and `last_user_sync_ts`. `last_active_ts` is now handled by `set_state()`, and `last_user_sync_ts` will be set at the finish of the context manager, so is unneeded here.

File: synapse/handlers/presence.py [hunk 2]
```diff
@@ -1866,7 +1862,7 @@ def handle_timeout(
     """Checks the presence of the user to see if any of the timers have elapsed
 
     Args:
-        state
+        state: the current state of the user, not the state at the setting of the timer
```
Comment (realtyem) [inline synapse/handlers/presence.py:1865]: This was just a driveby fix

File: synapse/handlers/presence.py [hunk 3]
```diff
@@ -1255,9 +1250,10 @@ async def set_state(
         if not ignore_status_msg:
             new_fields["status_msg"] = status_msg
 
-        if presence == PresenceState.ONLINE or (
-            presence == PresenceState.BUSY and self._busy_presence_enabled
-        ):
+        if (
+            prev_state.state != PresenceState.ONLINE
+            and presence == PresenceState.ONLINE
+        ) or (presence == PresenceState.BUSY and self._busy_presence_enabled):
```
Comment (realtyem) [inline synapse/handlers/presence.py:1256]: This was the meat and potatoes of the fix, if we are already `online`, we don't have to bump `last_active_ts`

File: tests/handlers/test_presence.py [hunk 4]
```diff
@@ -773,6 +773,32 @@ def test_set_presence_from_syncing_keeps_busy(
...
+        self.reactor.advance(2)
+        self.get_success(
+            self.presence_handler.user_syncing(user_id, True, PresenceState.ONLINE)
+        )
+
+        state = self.get_success(
+            self.presence_handler.get_state(UserID.from_string(user_id))
+        )
+        # last_active_ts should not have changed as no pro-active event has occurred
+        self.assertEqual(prev_state.last_active_ts, state.last_active_ts)
```
Comment (realtyem) [inline tests/handlers/test_presence.py:800]: FYI: on `develop` this fails with
```log
twisted.trial.unittest.FailTest: 2000 != 4000
```

File: synapse/handlers/presence.py [hunk 5]
```diff
@@ -1005,35 +1006,29 @@ async def user_syncing(
...
             # otherwise, set the new presence state & update the last sync time,
             # but don't update last_active_ts as this isn't an indication that
             # they've been active (even though it's probably been updated by
-            # set_state above)
-            else:
+            # set_state above).
+            # TODO: This will be removed in future work as it's
+            #  literally done again at the '_end' of this context manager and the
+            #  only time last_user_sync_ts matters is when processing timeouts in
+            #  handle_timeout().
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:1030]: Then can't we just remove it now rather than leaving a comment?

**Review by erikjohnston — COMMENTED** [review]
Otherwise seems good

File: synapse/handlers/presence.py (hunk 5 above)
Comment (realtyem) [reply synapse/handlers/presence.py:1030]: Not yet. And I'm on the fence based on further experiments about putting the other code block back in. Let's put this on hold for the moment. I think I have something to show you(out-of-band though).

**Comment by realtyem** [comment]
I'm going to close this for now, as there were unforeseen side-effects I'm not happy about. I will roll this into another pull request with more in it to abate those side-effects.

(Mainly, the side effects were that federation sending would ramp up and stay up for over an hour. I couldn't pin down exactly what was causing this, but fixing 'presence spam' seemed to make it stop and requires something broader than this one PR was scoped for.)

**Comment by realtyem** [comment]
Replaced with #15989
